File: main/1 - core/engine/modules/elementLibrary/circleWithOutline.js
this.circleWithOutline = function(_id,_name){
    const self = this;

    //attributes 
        //protected attributes
            const type = 'circleWithOutline'; 
            this.getType = function(){return type;}
            const id = _id; 
            this.getId = function(){return id;}

        //simple attributes
            this.name = _name;
            this.parent = undefined;
            this.dotFrame = false;
            this.extremities = { points:[], boundingBox:{bottomRight:{x:0, y:0}, topLeft:{x:0, y:0}} };
            let ignored = false;
            this.ignored = function(a){
                if(a==undefined){return ignored;}     
                ignored = a;
                dev.log.elementLibrary[type]('['+self.getAddress()+'].ignored(',a); //#development
                if(allowComputeExtremities){computeExtremities();}
            };
            let colour = {r:1,g:0,b:0,a:1};
            this.colour = function(a){
                if(a==undefined){return colour;}     
                colour = a;
                dev.log.elementLibrary[type]('['+self.getAddress()+'].colour(',a); //#development
            };
            let lineColour = {r:1,g:0,b:0,a:1};
            this.lineColour = function(a){
                if(a==undefined){return lineColour;}     
                lineColour = a;
                dev.log.elementLibrary[type]('['+self.getAddress()+'].lineColour(',a); //#development
            };
        
        //advanced use attributes
            let allowComputeExtremities = true;

        //addressing
            this.getAddress = function(){ return (this.parent != undefined ? this.parent.getAddress() : '') + '/' + this.name; };

        //attributes pertinent to extremity calculation
            let x = 0;
            let y = 0; 
            let angle = 0;
            let radius = 10;
            let detail = 25;
            let scale = 1;
            let thickness = 0;
            let isStatic = false;
            this.x = function(a){ 
                if(a==undefined){return x;}     
                x = a;     
                dev.log.elementLibrary[type]('['+self.getAddress()+'].x(',a); //#development
                if(allowComputeExtremities){computeExtremities();}
            };
            this.y = function(a){ 
                if(a==undefined){return y;}     
                y = a;
                dev.log.elementLibrary[type]('['+self.getAddress()+'].y(',a); //#development
                if(allowComputeExtremities){computeExtremities();}
            };
            this.radius = function(a){ 
                if(a==undefined){return radius;} 
                radius = a;
                dev.log.elementLibrary[type]('['+self.getAddress()+'].radius(',a); //#development
                if(allowComputeExtremities){computeExtremities();}
            };
            this.detail = function(a){ 
                if(a==undefined){return detail;}
                detail = a;
                dev.log.elementLibrary[type]('['+self.getAddress()+'].detail(',a); //#development
                calculateCirclePoints();
                if(allowComputeExtremities){computeExtremities();}
            };
            this.scale = function(a){ 
                if(a==undefined){return scale;} 
                scale = a;
                dev.log.elementLibrary[type]('['+self.getAddress()+'].scale(',a); //#development
                if(allowComputeExtremities){computeExtremities();}
            };
            this.thickness = function(a){ 
                if(a==undefined){return thickness;} 
                thickness = a;
                dev.log.elementLibrary[type]('['+self.getAddress()+'].thickness(',a); //#development
                if(allowComputeExtremities){computeExtremities();}
            };
            this.static = function(a){
                if(a==undefined){return isStatic;}  
                isStatic = a;  
                dev.log.elementLibrary[type]('['+self.getAddress()+'].static(',a); //#development
                if(allowComputeExtremities){computeExtremities();}
            };

        //unifiedAttribute
            this.unifiedAttribute = function(attributes){
                if(attributes==undefined){ return { ignored:ignored, colour:colour, lineColour:lineColour, x:x, y:y, angle:angle, radius:radius, detail:detail, scale:scale, thickness:thickness, static:isStatic}; } 
                dev.log.elementLibrary[type]('['+self.getAddress()+'].unifiedAttribute(',attributes); //#development

                allowComputeExtremities = false;
                Object.keys(attributes).forEach(key => {
                    dev.log.elementLibrary[type]('['+self.getAddress()+'].unifiedAttribute -> updating "'+key+'" to '+JSON.stringify(attributes[key])); //#development
                    try{
                        self[key](attributes[key]);
                    }catch(err){
                        console.warn(type,id,self.getAddress(),'.unifiedAttribute -> unknown attribute "'+key+'" which was being set to "'+JSON.stringify(attributes[key])+'"');
                    }
                });
                allowComputeExtremities = true;

                computeExtremities();
            };

    //webGL rendering functions
        let points = []; 
        let pointsChanged = true;
        function calculateCirclePoints(){
            points = [];
            //outline
                for(let a = 0; a < detail; a++){
                    points.push(0,0);
                    points.push( Math.sin( 2*Math.PI * (a/detail) ), Math.cos( 2*Math.PI * (a/detail) ) );
                    points.push( Math.sin( 2*Math.PI * ((a+1)/detail) ), Math.cos( 2*Math.PI * ((a+1)/detail) ) );
                }
            //main circle
                for(let a = 0; a < detail; a++){
                    points.push(0,0);
                    points.push( Math.sin( 2*Math.PI * (a/detail) ), Math.cos( 2*Math.PI * (a/detail) ) );
                    points.push( Math.sin( 2*Math.PI * ((a+1)/detail) ), Math.cos( 2*Math.PI * ((a+1)/detail) ) );
                }
            pointsChanged = true;
        }
        calculateCirclePoints();

        const vertexShaderSource = 
            '#version 300 es' + library.glsl.geometry + `
            //index
                in lowp float index;
            
            //constants
                in vec2 point;

            //variables
                struct location{
                    vec2 xy;
                    float scale;
                    float angle;
                };
                uniform location adjust;

                uniform vec2 resolution;
                uniform float radius;
                uniform float thickness;
                uniform vec4 colour;
                uniform vec4 lineColour;
                uniform lowp float indexParting;
        
            //varyings
                out vec4 activeColour;

            void main(){    
                //adjust points by radius and xy adjust
                    float tmpRadius = radius + (thickness/2.0) * (index < indexParting ? 1.0 : -1.0);
                    vec2 P = cartesianAngleAdjust(point*tmpRadius*adjust.scale, -adjust.angle) + adjust.xy;

                //select colour
                    activeColour = index >= indexParting ? colour : lineColour;

                //convert from unit space to clipspace
                    gl_Position = vec4( (((P / resolution) * 2.0) - 1.0) * vec2(1, -1), 0, 1 );
            }
        `;
        const fragmentShaderSource = `#version 300 es
            precision mediump float;
            out vec4 outputColour;
            in vec4 activeColour;
                                                                        
            void main(){
                outputColour = activeColour;
            }
        `;
        const point = { buffer:undefined, attributeLocation:undefined };
        const index = { buffer:undefined, attributeLocation:undefined };
        let uniformLocations;
        function updateGLAttributes(context,adjust){
            dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes(',context,adjust); //#development

            //buffers
                //points
                    if(point.buffer == undefined || pointsChanged){
                        dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes -> creating point.buffer...'); //#development
                        point.attributeLocation = context.getAttribLocation(program, "point");
                        point.buffer = context.createBuffer();
                        context.enableVertexAttribArray(point.attributeLocation);
                        context.bindBuffer(context.ARRAY_BUFFER, point.buffer); 
                        context.vertexAttribPointer( point.attributeLocation, 2, context.FLOAT,false, 0, 0 );
                        context.bufferData(context.ARRAY_BUFFER, new Float32Array(points), context.STATIC_DRAW);
                        pointsChanged = false;
                    }else{
                        dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes -> updating point.buffer...'); //#development
                        context.bindBuffer(context.ARRAY_BUFFER, point.buffer); 
                        context.vertexAttribPointer( point.attributeLocation, 2, context.FLOAT,false, 0, 0 );
                }
                //index
                    if(index.buffer == undefined || pointsChanged){
                        dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes -> creating index.buffer...'); //#development
                        index.attributeLocation = context.getAttribLocation(program, "index");
                        index.buffer = context.createBuffer();
                        context.enableVertexAttribArray(index.attributeLocation);
                        context.bindBuffer(context.ARRAY_BUFFER, index.buffer); 
                        context.vertexAttribPointer( index.attributeLocation, 1, context.FLOAT, false, 0, 0 );
                        context.bufferData(context.ARRAY_BUFFER, new Float32Array(Array.apply(null, {length:points.length/2}).map(Number.call, Number)), context.STATIC_DRAW);
                    }else{
                        dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes -> updating index.buffer...'); //#development
                        context.bindBuffer(context.ARRAY_BUFFER, index.buffer);
                        context.vertexAttribPointer( index.attributeLocation, 1, context.FLOAT, false, 0, 0 );
                    }

            //uniforms
                if(uniformLocations == undefined){
                    dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes -> defining uniformLocations...'); //#development
                    uniformLocations = {
                        "adjust.xy": context.getUniformLocation(program, "adjust.xy"),
                        "adjust.scale": context.getUniformLocation(program, "adjust.scale"),
                        "adjust.angle": context.getUniformLocation(program, "adjust.angle"),
                        "resolution": context.getUniformLocation(program, "resolution"),
                        "radius": context.getUniformLocation(program, "radius"),
                        "thickness": context.getUniformLocation(program, "thickness"),
                        "colour": context.getUniformLocation(program, "colour"),
                        "indexParting": context.getUniformLocation(program, "indexParting"),
                        "lineColour": context.getUniformLocation(program, "lineColour"),
                    };
                }

                dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes -> adjust.x:'+adjust.x+' adjust.y:'+adjust.y); //#development
                dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes -> adjust.scale:'+adjust.scale); //#development
                dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes -> adjust.angle:'+adjust.angle); //#development
                dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes -> canvas.width:'+context.canvas.width+' canvas.height:'+context.canvas.height); //#development
                dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes -> radius:'+radius); //#development
                dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes -> thickness:'+thickness); //#development
                dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes -> colour:',colour); //#development
                dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes -> lineColour:',lineColour); //#development
                dev.log.elementLibrary[type]('['+self.getAddress()+']::updateGLAttributes -> indexParting:'+points.length/4); //#development
                context.uniform2f(uniformLocations["adjust.xy"], adjust.x, adjust.y);
                context.uniform1f(uniformLocations["adjust.scale"], adjust.scale);
                context.uniform1f(uniformLocations["adjust.angle"], adjust.angle);
                context.uniform2f(uniformLocations["resolution"], context.canvas.width, context.canvas.height);
                context.uniform1f(uniformLocations["radius"], radius);
                context.uniform1f(uniformLocations["thickness"], thickness);
                context.uniform4f(uniformLocations["colour"], colour.r*colour.a, colour.g*colour.a, colour.b*colour.a, colour.a);
                context.uniform4f(uniformLocations["lineColour"], lineColour.r*lineColour.a, lineColour.g*lineColour.a, lineColour.b*lineColour.a, lineColour.a);
                context.uniform1f(uniformLocations["indexParting"], points.length/4);
        }
        let program;
        function activateGLRender(context,adjust){
            dev.log.elementLibrary[type]('['+self.getAddress()+']::activateGLRender(',context,adjust); //#development
            if(program == undefined){ program = render.produceProgram(self.getType(), vertexShaderSource, fragmentShaderSource); }

            context.useProgram(program);
            updateGLAttributes(context,adjust);
            context.drawArrays(context.TRIANGLES, 0, points.length/2);
        }

    //extremities
        function computeExtremities(informParent=true,offset){
            dev.log.elementLibrary[type]('['+self.getAddress()+']::computeExtremities(',informParent,offset); //#development

            //get offset from parent, if one isn't provided
                if(offset == undefined){ offset = self.parent && !self.static() ? self.parent.getOffset() : {x:0,y:0,scale:1,angle:0}; }
            //calculate adjusted offset based on the offset
                const point = library.math.cartesianAngleAdjust(x,y,offset.angle);
                const adjusted = { 
                    x: point.x*offset.scale + offset.x,
                    y: point.y*offset.scale + offset.y,
                    scale: offset.scale*scale,
                    angle: -(offset.angle + angle),
                };
            //calculate points based on the adjusted offset
                self.extremities.points = [];
                for(let a = 0; a < points.length; a+=2){
                    self.extremities.points.push({
                        x: (points[a]   * radius * adjusted.scale) + adjusted.x,
                        y: (points[a+1] * radius * adjusted.scale) + adjusted.y,
                    });
                }
                self.extremities.boundingBox = library.math.boundingBoxFromPoints(self.extremities.points);
            //if told to do so, inform parent (if there is one) that extremities have changed
                if(informParent){ if(self.parent){self.parent.updateExtremities();} }
        }
        this.computeExtremities = computeExtremities;

    //lead render
        function drawDotFrame(){
            //draw shape extremity points
                self.extremities.points.forEach(a => render.drawDot(a.x,a.y));
            //draw bounding box top left and bottom right points
                render.drawDot(self.extremities.boundingBox.topLeft.x,self.extremities.boundingBox.topLeft.y,2,{r:0,g:0,b:1,a:1});
                render.drawDot(self.extremities.boundingBox.bottomRight.x,self.extremities.boundingBox.bottomRight.y,2,{r:0,g:0,b:1,a:1});
        };
        this.render = function(context,offset={x:0,y:0,scale:1,angle:0}){
            dev.log.elementLibrary[type]('['+self.getAddress()+'].render(',context,offset); //#development
            //combine offset with shape's position, angle and scale to produce adjust value for render
                const point = library.math.cartesianAngleAdjust(x,y,offset.angle);
                const adjust = { 
                    x: point.x*offset.scale + offset.x,
                    y: point.y*offset.scale + offset.y,
                    scale: offset.scale*scale,
                    angle: -(offset.angle + angle),
                };

            //activate shape render code
                activateGLRender(context,adjust);

            //if requested; draw dot frame
                if(self.dotFrame){drawDotFrame();}
        };

    //info dump
        this._dump = function(){
            report.info(self.getAddress(),'._dump()');
            report.info(self.getAddress(),'._dump -> id: '+id);
            report.info(self.getAddress(),'._dump -> type: '+type);
            report.info(self.getAddress(),'._dump -> name: '+self.name);
            report.info(self.getAddress(),'._dump -> address: '+self.getAddress());
            report.info(self.getAddress(),'._dump -> parent: '+JSON.stringify(self.parent));
            report.info(self.getAddress(),'._dump -> dotFrame: '+self.dotFrame);
            report.info(self.getAddress(),'._dump -> extremities: '+JSON.stringify(self.extremities));
            report.info(self.getAddress(),'._dump -> ignored: '+ignored);
            report.info(self.getAddress(),'._dump -> colour: '+JSON.stringify(colour));
            report.info(self.getAddress(),'._dump -> lineColour: '+JSON.stringify(lineColour));
            report.info(self.getAddress(),'._dump -> x: '+x);
            report.info(self.getAddress(),'._dump -> y: '+y);
            report.info(self.getAddress(),'._dump -> radius: '+radius);
            report.info(self.getAddress(),'._dump -> detail: '+detail);
            report.info(self.getAddress(),'._dump -> scale: '+scale);
            report.info(self.getAddress(),'._dump -> thickness: '+thickness);
            report.info(self.getAddress(),'._dump -> static: '+self.static());
        };
    
    //interface
        this.interface = new function(){
            this.ignored = self.ignored;
            this.colour = self.colour;
            this.lineColour = self.lineColour;
            this.x = self.x;
            this.y = self.y;
            this.radius = self.radius;
            this.scale = self.scale;
            this.thickness = self.thickness;
            this.static = self.static;
            this.unifiedAttribute = self.unifiedAttribute;
            this.getAddress = self.getAddress;
            this._dump = self._dump;
        };
};

File: trash/development2/src/b_element/element/circleWithOutline.rs
//rust
    use std::rc::Weak;
    use std::cell::RefCell;
    use std::f32::consts::PI;

//wasm
    use web_sys::WebGl2RenderingContext;

//core
    use crate::a_library::{
        data_type::{
            Colour,
            Offset,
            Polygon,
            ElementType,
        },
        structure::{
            WebGl2programConglomerateManager,
            ImageRequester,
            FontRequester,
        },
        wasm::from_js_sys::{
            get_value_from_object__u32,
            get_value_from_object__f32,
            get_value_from_object__colour,
        },
    };
    use super::super::element::ElementTrait;








static VAO_NAMES: [&'static str; 2] = [
    "index",
    "point",
];
// static VAO_INDEX: [f32; ?] = [
//     - generated -
// ];
// static VAO_POINTS: [f32; ?] = [
//     - generated -
// ];
static VAO_GROUPING_SIZE: [i32; 2] = [
    1,
    2,
];
static VERTEX_SHADER_SOURCE: &'static str = "#version 300 es
    #define PI 3.141592653589793

    vec2 cartesianAngleAdjust(vec2 xy, float angle){
        if(angle == 0.0 || mod(angle,PI*2.0) == 0.0){ return xy; }
        return vec2( xy.x*cos(angle) - xy.y*sin(angle), xy.y*cos(angle) + xy.x*sin(angle) ); 
    }

    //index
        in lowp float index;

    //constants
        in vec2 point;

    //variables
        struct location{
            vec2 xy;
            float scale;
            float angle;
        };
        uniform location adjust;

        uniform vec2 resolution;
        uniform float radius;
        uniform float thickness;
        uniform vec4 colour;
        uniform vec4 line_colour;
        uniform lowp float indexParting;

    //varyings
        out vec4 activeColour;

    void main(){    
        //adjust points by radius and xy adjust
            float tmpRadius = radius + (thickness/2.0) * (index < indexParting ? 1.0 : -1.0);
            vec2 P = cartesianAngleAdjust(point*tmpRadius*adjust.scale, -adjust.angle) + adjust.xy;

        //select colour
            activeColour = index >= indexParting ? colour : line_colour;

        //convert from unit space to clipspace
            gl_Position = vec4( (((P / resolution) * 2.0) - 1.0) * vec2(1, -1), 0, 1 );
    }
";
static FRAGMENT_SHADER_SOURCE: &'static str = "#version 300 es
    precision mediump float;
    out vec4 outputColour;
    in vec4 activeColour;
                                                                
    void main(){
        outputColour = activeColour;
    }
";
static UNIFORMS_TO_FIND: [&'static str; 9] = [
    "adjust.xy",
    "adjust.scale",
    "adjust.angle",
    "resolution",
    "radius",
    "thickness",
    "colour",
    "line_colour",
    "indexParting",
];

pub struct CircleWithOutline {
    //hierarchy and identity
        element_type: ElementType, 
        id: usize,
        name: String, 
        self_reference: Option<Weak<RefCell<dyn ElementTrait>>>,
        parent: Option<Weak<RefCell<dyn ElementTrait>>>,

    //attributes
        //pertinent to extremity calculation
            x: f32,
            y: f32,
            angle: f32,
            scale: f32,
            radius: f32,
            detail: u32,
            thickness: f32,
        //other
            ignored: bool,
            colour: Colour,
            line_colour: Colour,

    //switches
        dot_frame: bool,
        allow_compute_extremities: bool,

    //computed values
        extremities: Polygon,
        cached_offset: Offset,
        vao_id: Option<usize>,
        vao_index: Vec<f32>,
        vao_points: Vec<f32>,
        points_changed: bool,
}
impl CircleWithOutline {
    pub fn new(id:usize, name:String) -> CircleWithOutline {
        CircleWithOutline {
            element_type: ElementType::CircleWithOutline,
            id: id,
            name: name,
            self_reference: None,
            parent: None,

            x: 0.0,
            y: 0.0,
            angle: 0.0,
            scale: 1.0,
            radius: 10.0,
            detail: 25,
            thickness: 0.0,

            ignored: false,
            colour: Colour::new(1.0,0.0,0.0,1.0),
            line_colour: Colour::new(1.0,0.0,0.0,1.0),

            dot_frame: false,
            allow_compute_extremities: true,
            
            extremities: Polygon::new_empty(),
            cached_offset: Offset::new_default(),
            vao_id: None,
            vao_index: (0..25*2*6).map(|a| a as f32).collect(),
            vao_points: CircleWithOutline::compute_points(25),
            points_changed: false,
        }
    }

    //attributes
        //pertinent to extremity calculation
            //radius
                pub fn get_radius(&self) -> &f32 { &self.radius }
                pub fn set_radius(&mut self, new:f32) { self.radius = new; self.compute_extremities(true, None, None); }
            //detail
                pub fn get_detail(&self) -> &u32 { &self.detail }
                pub fn set_detail(&mut self, new:u32) { self.detail = new; self.calculate_circle_points(); self.compute_extremities(true, None, None); }
            //thickness
                pub fn get_thickness(&self) -> &f32 { &self.thickness }
                pub fn set_thickness(&mut self, new:f32) { self.thickness = new; self.compute_extremities(true, None, None); }
        //other
            pub fn get_colour(&self) -> &Colour { &self.colour }
            pub fn set_colour(&mut self, new:Colour) { self.colour = new; }
            pub fn get_line_colour(&self) -> &Colour { &self.line_colour }
            pub fn set_line_colour(&mut self, new:Colour) { self.line_colour = new; }
        //unified attribute
            pub fn set_unified_attribute(
                &mut self,
                x: Option<f32>,
                y: Option<f32>,
                angle: Option<f32>,
                scale: Option<f32>,
                radius: Option<f32>,
                detail: Option<u32>,
                thickness: Option<f32>,
                colour: Option<Colour>,
                line_colour: Option<Colour>,
            ) {
                if let Some(x) = x { self.x = x; }
                if let Some(y) = y { self.y = y; }
                if let Some(angle) = angle { self.angle = angle; }
                if let Some(scale) = scale { self.scale = scale; }
                if let Some(radius) = radius { self.radius = radius; }
                if let Some(detail) = detail { self.detail = detail; self.calculate_circle_points(); }
                if let Some(thickness) = thickness { self.thickness = thickness; }
                if let Some(colour) = colour { self.colour = colour; }
                if let Some(line_colour) = line_colour { self.line_colour = line_colour; }
                self.compute_extremities(true, None, None);
            }

    //webGL rendering functions
        fn compute_points(detail:u32) -> Vec<f32> {
            let mut vao_points:Vec<f32> = vec![];
            let d = detail as f32;

            //outline
                for index in 0..detail {
                    let i = index as f32;
                    vao_points.extend_from_slice( &[0.0,0.0] );
                    vao_points.extend_from_slice( &[ ( 2.0*PI * (i/d) ).sin(), ( 2.0*PI * (i/d) ).cos() ] );
                    vao_points.extend_from_slice( &[ ( 2.0*PI * ((i+1.0)/d) ).sin(), ( 2.0*PI * ((i+1.0)/d) ).cos() ] );
                }

            //main circle
                for index in 0..detail {
                    let i = index as f32;
                    vao_points.extend_from_slice( &[0.0,0.0] );
                    vao_points.extend_from_slice( &[ ( 2.0*PI * (i/d) ).sin(), ( 2.0*PI * (i/d) ).cos() ] );
                    vao_points.extend_from_slice( &[ ( 2.0*PI * ((i+1.0)/d) ).sin(), ( 2.0*PI * ((i+1.0)/d) ).cos() ] );
                }

            vao_points
        }
        fn calculate_circle_points(&mut self) {
            self.vao_points = CircleWithOutline::compute_points(self.detail);
            self.vao_index = (0..self.vao_points.len()).map(|a| a as f32).collect();
            self.points_changed = true;
        }
}
impl ElementTrait for CircleWithOutline {
    //trait requirements
        //hierarchy and identity
            fn get_element_type(&self) -> ElementType { self.element_type }
            fn get_id(&self) -> usize { self.id }
            fn get_name(&self) -> &String{ &self.name }
            fn set_name(&mut self, new:String) { self.name = new; }
            fn set_self_reference(&mut self, new:Weak<RefCell<dyn ElementTrait>>) { self.self_reference = Some(new); }
            fn get_parent(&self) -> &Option<Weak<RefCell<dyn ElementTrait>>> { &self.parent }
            fn set_parent(&mut self, new:Weak<RefCell<dyn ElementTrait>>) { self.parent = Some(new); }
            fn clear_parent(&mut self) { self.parent = None; }

        //position
            //x
                fn get_x(&self) -> f32 { self.x }
                fn set_x(&mut self, new:f32){ self.x = new; self.compute_extremities(true, None, None); }
            //y
                fn get_y(&self) -> f32 { self.y }
                fn set_y(&mut self, new:f32){ self.y = new; self.compute_extremities(true, None, None); }
            //angle
                fn get_angle(&self) -> f32 { self.angle }
                fn set_angle(&mut self, new:f32) { self.angle = new; self.compute_extremities(true, None, None); }
            //scale
                fn get_scale(&self) -> f32 { self.scale }
                fn set_scale(&mut self, new:f32) { self.scale = new; self.compute_extremities(true, None, None); }

        //other
            fn get_ignored(&self) -> bool { self.ignored }
            fn set_ignored(&mut self, new:bool) { self.ignored = new; }

        //offset
            fn get_cached_offset(&self) -> &Offset { &self.cached_offset }
            fn get_cached_offset_mut(&mut self) -> &mut Offset { &mut self.cached_offset }
            fn set_cached_offset(&mut self, new:Offset) { self.cached_offset = new; }

        //extremities
            fn get_allow_compute_extremities(&self) -> bool { self.allow_compute_extremities }
            fn set_allow_compute_extremities(&mut self, new:bool) { self.allow_compute_extremities = new; }
            fn get_extremities(&self) -> &Polygon { &self.extremities }
            fn __set_extremities(&mut self, new:Polygon) { self.extremities = new; }

        //render
            fn get_dot_frame(&self) -> bool { self.dot_frame }
            fn set_dot_frame(&mut self, new:bool) { self.dot_frame = new; }

    //element specific
        //casting
            fn as_circleWithOutline(&self) -> Option<&CircleWithOutline> { Some(self) }
            fn as_circleWithOutline_mut(&mut self) -> Option<&mut CircleWithOutline> { Some(self) }

        //universal attribute
            fn set_unified_attribute_from_js_sys_object(&mut self, unified_attribute:js_sys::Object, _font_requester:&RefCell<FontRequester>, _worker:&web_sys::Worker) {
                self.set_unified_attribute(
                    get_value_from_object__f32("x", &unified_attribute, true),
                    get_value_from_object__f32("y", &unified_attribute, true),
                    get_value_from_object__f32("angle", &unified_attribute, true),
                    get_value_from_object__f32("scale", &unified_attribute, true),
                    get_value_from_object__f32("radius", &unified_attribute, true),
                    get_value_from_object__u32("detail", &unified_attribute, true),
                    get_value_from_object__f32("thickness", &unified_attribute, true),
                    get_value_from_object__colour("colour", &unified_attribute, true),
                    get_value_from_object__colour("lineColour", &unified_attribute, true),
                );
            }

        //extremities
            fn get_length_of_points_for_extremity_calculation(&self) -> usize {
                self.vao_points.len()/2
            }
            fn get_point_for_extremity_calculation(&self, index:usize) -> Option<(f32,f32)> {
                if index >= self.vao_points.len()/2 {
                    return None;
                }

                Some( (
                    self.radius * self.vao_points[index],
                    self.radius * self.vao_points[index+1],
                ) )
            }

        //render
            fn activate_webGL_render(
                &mut self, 
                offset: Option<&Offset>,
                context: &WebGl2RenderingContext, 
                web_gl2_program_conglomerate_manager: &mut WebGl2programConglomerateManager,
                _image_requester: &mut ImageRequester,
                resolution: &(u32, u32),
            ) {
                //vao
                    if self.vao_id.is_none() {
                        self.vao_id = Some( web_gl2_program_conglomerate_manager.generate_new_VAO_id(&self.element_type) );
                    }

                    if !web_gl2_program_conglomerate_manager.program_is_loaded(self.element_type) {
                        self.points_changed = false;
                    } else if self.points_changed {
                        web_gl2_program_conglomerate_manager.update_VAO(
                            &context,
                            self.element_type,
                            self.vao_id.unwrap(),
                            &VAO_NAMES,
                            &[&self.vao_index, &self.vao_points],
                            &VAO_GROUPING_SIZE,
                        );
                        self.points_changed = false;
                    }

                //load program
                    web_gl2_program_conglomerate_manager.load_program(
                        &context,
                        self.element_type,
                        &VERTEX_SHADER_SOURCE,
                        &FRAGMENT_SHADER_SOURCE,
                        self.vao_id.unwrap(),
                        &VAO_NAMES,
                        &[&self.vao_index, &self.vao_points],
                        &VAO_GROUPING_SIZE,
                        &UNIFORMS_TO_FIND,
                        None,
                    );

                //determine which offset to use
                    let working_offset = match offset {
                        None => self.get_cached_offset(),
                        Some(offset) => { offset },
                    };
                    
                //load uniforms
                    let uniforms = web_gl2_program_conglomerate_manager.get_uniform_locations(self.element_type).unwrap();
                    context.uniform2f(Some(&uniforms[0]), working_offset.get_x(), working_offset.get_y());
                    context.uniform1f(Some(&uniforms[1]), working_offset.get_scale());
                    context.uniform1f(Some(&uniforms[2]), working_offset.get_angle());
                    context.uniform2f(Some(&uniforms[3]), resolution.0 as f32, resolution.1 as f32);
                    context.uniform1f(Some(&uniforms[4]), self.radius);
                    context.uniform1f(Some(&uniforms[5]), self.thickness);
                    context.uniform4f(Some(&uniforms[6]), self.colour.premultiplied_r(), self.colour.premultiplied_g(), self.colour.premultiplied_b(), self.colour.a());
                    context.uniform4f(Some(&uniforms[7]), self.line_colour.premultiplied_r(), self.line_colour.premultiplied_g(), self.line_colour.premultiplied_b(), self.line_colour.a());
                    context.uniform1f(Some(&uniforms[8]), (self.vao_index.len()/4) as f32);

                //activate draw
                    context.draw_arrays(WebGl2RenderingContext::TRIANGLES, 0, (self.vao_points.len()/2) as i32);
            }

        //info/dump
            fn _specific_info(&self) -> String {
                format!(
                    "radius:{}, detail:{}, thickness:{}, colour:{}, line_colour:{}",
                    self.radius,
                    self.detail,
                    self.thickness,
                    self.colour, 
                    self.line_colour,
                )
            }
}